elyra/validation/validate.py

```diff
@@ -401,44 +423,57 @@ def _validate_pipeline_graph(self, pipeline: Dict, response):
         pipeline_id = pipeline_json['pipelines'][0]['id']
 
-        node_graph = []
+        graph = nx.DiGraph()
         for node in node_list:
+            graph.add_node(node['id'])
             if 'links' in node['inputs'][0]:
                 for link in node['inputs'][0]['links']:
-                    node_graph.append((node['id'], link['node_id_ref']))
-            else:
-                node_graph.append((node['id'], ""))
-
-        graph = nx.DiGraph(node_graph)
-
-        for node in node_list:
-            graph.add_node(node['id'])
+                    graph.add_edge(link['node_id_ref'], node['id'])
 
         for isolate in nx.isolates(graph):
-            response.add_message(severity='2',
+            response.add_message(severity=ValidationSeverity.Warning,
                                  message_type="singletonReference",
                                  message="This node is not connected to any part of the pipeline",
                                  data={"nodeID": isolate,
                                        "pipelineID": pipeline_id})
 
-        cycles_detected = nx.find_cycle(graph)
+        cycles_detected = nx.simple_cycles(graph)
 
+        link_dict_table = {}
+        cycle_counter = 1
         for cycle in cycles_detected:
-            if len(cycle) == 1:
-                for node in node_list:
-                    for link in node['inputs'][0]['links']:
-                        if cycle[0] == link['node_id_ref']:
-                            response.add_message(severity=ValidationSeverity.Error,
-                                                 message_type="circularReference",
-                                                 message="A cycle was found within this pipeline",
-                                                 data={"nodeID": node['id'],
-                                                       "pipelineID": pipeline_id})
-            else:
-                for node in node_list:
-                    for link in node['inputs'][0]['links']:
-                        if cycle[0] == link['node_id_ref'] and cycle[1] == node['id']:
-                            response.add_message(severity=ValidationSeverity.Error,
-                                                 message_type="circularReference",
-                                                 message="A cycle was found within this pipeline",
-                                                 data={"nodeID": node['id'],
-                                                       "pipelineID": pipeline_id})
+            size_of_cycle = len(cycle)
+            for i in range(0, size_of_cycle):
+                if cycle_counter not in link_dict_table:
+                    link_dict_table[cycle_counter] = []
+                if size_of_cycle == 1:
+                    link_dict_table[cycle_counter].append(self._get_link_id(pipeline, cycle[i], cycle[i]))
+                elif i == size_of_cycle - 1:
```

Comment:
These first two clauses are redundant since a cycle of 1 will have `i` be the last node of the cycle.  It might be easier to read if you defined 'the last node' as requiring nodes (`i` and `0`), while all others use (`i` and `i+1`).